selfie.c

```diff
@@ -7632,366 +8817,325 @@ void fetch() {
+    return;
+  else if (is == ECALL)
+    undo_ecall();
+  else
+    undo_lui_addi_add_sub_mul_divu_remu_sltu_ld_jal_jalr();
+}
+
+void execute_debug() {
+  translate_to_assembler();
+
+  // assert: 1 <= is <= number of RISC-U instructions
+  if (is == ADDI){
+    print_addi_before();
+    do_addi();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == LD) {
+    print_ld_before();
+    print_ld_after(do_ld());
+  } else if (is == SD) {
+    print_sd_before();
+    print_sd_after(do_sd());
+  } else if (is == ADD) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_add();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == SUB) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_sub();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == MUL) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_mul();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == DIVU) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_divu();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == REMU) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_remu();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == SLTU) {
+    print_add_sub_mul_divu_remu_sltu_before();
+    do_sltu();
+    print_addi_add_sub_mul_divu_remu_sltu_after();
+  } else if (is == BEQ) {
+    print_beq_before();
+    do_beq();
+    print_beq_after();
+  } else if (is == JAL) {
+    print_jal_before();
+    do_jal();
+    print_jal_jalr_after();
+  } else if (is == JALR) {
+    print_jalr_before();
+    do_jalr();
+    print_jal_jalr_after();
+  } else if (is == LUI) {
+    print_lui_before();
+    do_lui();
+    print_lui_after();
+  } else if (is == ECALL) {
+    do_ecall();
 
-    exit(EXITCODE_UNKNOWNINSTRUCTION);
+    return;
   }
+
+  println();
+}
+
+void execute_symbolically() {
+  // assert: 1 <= is <= number of RISC-U instructions
+  if (is == ADDI) {
+    constrain_addi();
+    do_addi();
+  } else if (is == LD)
+    constrain_ld();
+  else if (is == SD)
+    constrain_sd();
+  else if (is == ADD) {
+    constrain_add_sub_mul_divu_remu_sltu("bvadd");
+    do_add();
+  } else if (is == SUB) {
+    constrain_add_sub_mul_divu_remu_sltu("bvsub");
+    do_sub();
+  } else if (is == MUL) {
+    constrain_add_sub_mul_divu_remu_sltu("bvmul");
+    do_mul();
+  } else if (is == DIVU) {
+    constrain_add_sub_mul_divu_remu_sltu("bvudiv");
+    do_divu();
+  } else if (is == REMU) {
+    constrain_add_sub_mul_divu_remu_sltu("bvurem");
+    do_remu();
+  } else if (is == SLTU) {
+    constrain_add_sub_mul_divu_remu_sltu("bvult");
+    zero_extend_sltu();
+    do_sltu();
+  } else if (is == BEQ)
+    constrain_beq();
+  else if (is == JAL)
+    do_jal();
+  else if (is == JALR) {
+    constrain_jalr();
+    do_jalr();
+  } else if (is == LUI) {
+    constrain_lui();
+    do_lui();
+  } else if (is == ECALL)
+    do_ecall();
+}
+
+void execute_translate() {
```

Comment:
I am assuming this is not execution, right? You only go through the code once and translate each instruction, right? If yes, I'd call this procedure `translate_to_x86_binary`.

Reply:
I am using the prefix `execute_` only for procedures that actually execute code in some way.